File: backend/tests/builders/feature_flag.py
"""Builders for the feature-flag wire shapes: what the admin page sends on PATCH.

`build()` returns the camelCase JSON mapping the API actually receives, so tests exercise the
aliases and the validators rather than a pre-parsed object. Defaults are the seeded
`result_screen_tone` flag, so a test states only what it changes.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from datetime import UTC, datetime

SEEDED_UPDATED_AT = datetime(2026, 9, 18, 8, 0, tzinfo=UTC)


def a_wire_variant() -> _WireVariantBuilder:
    return _WireVariantBuilder()


@dataclass(frozen=True)
class _WireVariantBuilder:
    key: str = "calm"
    weight: int = 50
    headline: str = "Known breaches"
    subheadline: str = "Here's the public record of data breaches."
    cta_label: str = "Protect me"
    tone: str = "calm"

    def with_key(self, key: str) -> _WireVariantBuilder:
        return replace(self, key=key)

    def with_weight(self, weight: int) -> _WireVariantBuilder:
        return replace(self, weight=weight)

    def with_cta_label(self, cta_label: str) -> _WireVariantBuilder:
        return replace(self, cta_label=cta_label)

    def with_tone(self, tone: str) -> _WireVariantBuilder:
        return replace(self, tone=tone)

    def build(self) -> dict[str, object]:
        return {
            "key": self.key,
            "weight": self.weight,
            "config": {
                "headline": self.headline,
                "subheadline": self.subheadline,
                "ctaLabel": self.cta_label,
                "tone": self.tone,
            },
        }


def _calm_and_urgent() -> tuple[_WireVariantBuilder, ...]:
    return (
        a_wire_variant().with_key("calm").with_weight(50),
        a_wire_variant().with_key("urgent").with_weight(50).with_tone("urgent"),
    )


def a_feature_flag_update() -> _FeatureFlagUpdateBuilder:
    return _FeatureFlagUpdateBuilder()


@dataclass(frozen=True)
class _FeatureFlagUpdateBuilder:
    description: str = "Tone of the result screen."
    is_enabled: bool = True
    variants: tuple[_WireVariantBuilder, ...] = field(default_factory=_calm_and_urgent)
    updated_at: datetime = SEEDED_UPDATED_AT

    def with_variants(self, *variants: _WireVariantBuilder) -> _FeatureFlagUpdateBuilder:
        return replace(self, variants=variants)

    def disabled(self) -> _FeatureFlagUpdateBuilder:
        return replace(self, is_enabled=False)

    def with_updated_at(self, updated_at: datetime) -> _FeatureFlagUpdateBuilder:
        return replace(self, updated_at=updated_at)

    def build(self) -> dict[str, object]:
        return {
            "description": self.description,
            "isEnabled": self.is_enabled,
            "variants": [variant.build() for variant in self.variants],
            "updatedAt": self.updated_at.isoformat(),
        }

File: frontend/src/testkit/builders/featureFlag.ts
// Builders for the feature-flag wire shapes. Defaults are the seeded `result_screen_tone` flag.
//
// `with*` reassigns `this.state` to a fresh object and never mutates it in place: `build()`
// hands back the live object, so an in-place mutation would silently rewrite a DTO a previous
// `build()` already returned.

import type { FeatureFlagDTO, FeatureFlagVariantDTO } from '~/models/featureFlag';

const CALM: FeatureFlagVariantDTO = {
    key: 'calm',
    weight: 50,
    config: {
        headline: 'Known breaches',
        subheadline: "Here's the public record of data breaches.",
        ctaLabel: 'Protect me',
        tone: 'calm',
    },
};

const URGENT: FeatureFlagVariantDTO = {
    key: 'urgent',
    weight: 50,
    config: {
        headline: "You're exposed!",
        subheadline: '17.7B accounts have leaked. Yours could be among them.',
        ctaLabel: 'Protect me now',
        tone: 'urgent',
    },
};

const RESULT_SCREEN_TONE: FeatureFlagDTO = {
    key: 'result_screen_tone',
    description: 'Tone of the result screen: calm framing vs urgent framing of the same data.',
    isEnabled: true,
    variants: [CALM, URGENT],
    createdAt: '2026-09-18T08:00:00Z',
    updatedAt: '2026-09-18T08:00:00.123456Z',
};

class FeatureFlagDTOBuilder {
    private state: FeatureFlagDTO = { ...RESULT_SCREEN_TONE };

    withKey(key: string): this {
        this.state = { ...this.state, key };

        return this;
    }

    withUpdatedAt(updatedAt: string): this {
        this.state = { ...this.state, updatedAt };

        return this;
    }

    disabled(): this {
        this.state = { ...this.state, isEnabled: false };

        return this;
    }

    withUrgentCtaLabel(ctaLabel: string): this {
        this.state = {
            ...this.state,
            variants: this.state.variants.map((variant) =>
                variant.key === 'urgent' ? { ...variant, config: { ...variant.config, ctaLabel } } : variant,
            ),
        };

        return this;
    }

    build(): FeatureFlagDTO {
        return this.state;
    }
}

export const aFeatureFlagDTO = (): FeatureFlagDTOBuilder => new FeatureFlagDTOBuilder();
